db-init.sql

```diff
@@ -0,0 +1,166 @@
+	phoneNumber VARCHAR(20),
+	phoneArea VARCHAR(20),
+	phoneCountry VARCHAR(20),
+	companyName VARCHAR(255),
+	personCategoryId SERIAL REFERENCES PersonCategories(personCategoryId)
+);
+
+-- PersonAttributes
+	-- the values of the attributes that a specific person has
+
+CREATE TABLE PersonAttributes (
+	personUuid UUID REFERENCES Persons(personUuid) NOT NULL,
+	attrId SERIAL REFERENCES PersonAttributeTypes(attrId) NOT NULL,
+	attrValue VARCHAR(255)
+);
+
+-- MoneyTransactions
+
+CREATE TABLE MoneyTransactions (
+	moneyTransactionUuid UUID PRIMARY KEY NOT NULL DEFAULT uuid_generate_v4(),
+	datetime TIMESTAMP WITH TIME ZONE NOT NULL,
+	toPersonUuid UUID REFERENCES Persons(personUuid) NOT NULL,
+	fromPersonUuid UUID REFERENCES Persons(personUuid) NOT NULL,
+	amount MONEY NOT NULL
+);
+
+-- ProductTypes
+	-- all the different kinds of products
+
+CREATE TABLE ProductTypes (
+	productTypeId SERIAL PRIMARY KEY NOT NULL,
+	productName VARCHAR(255) NOT NULL,
+	productUnits VARCHAR(255) NOT NULL
+);
+
+	CREATE UNIQUE INDEX productNames_lower ON ProductTypes(lower(productName));
+
+	INSERT INTO ProductTypes VALUES (0, 'milk', 'litres');
+
+-- ProductTransactions
+
+CREATE TABLE ProductTransactions (
+	productTransactionUuid UUID PRIMARY KEY NOT NULL DEFAULT uuid_generate_v4(),
+	datetime TIMESTAMP WITH TIME ZONE NOT NULL,
+	toPersonUuid UUID REFERENCES Persons(personUuid) NOT NULL,
+	fromPersonUuid UUID REFERENCES Persons(personUuid) NOT NULL,
+	productTypeId SERIAL REFERENCES ProductTypes(productTypeId) NOT NULL,
+	amountOfProduct REAL NOT NULL,
+	costPerUnit MONEY NOT NULL
+);
+
+-- ProductTypeTransactionAttributes
+	-- the unique kinds of attributes a transaction for a type of product has
+	-- example: we track the milk density and milk temperature for all milk type transactions
+
+CREATE TABLE ProductTypeTransactionAttributes (
+	productTypeId SERIAL REFERENCES ProductTypes(productTypeId) NOT NULL,
+	attrId SERIAL PRIMARY KEY NOT NULL,
+	attrName VARCHAR(255) NOT NULL,
+	UNIQUE(productTypeId, attrId) -- every product type, may not have duplications of types of attributes
+);
+
+	CREATE UNIQUE INDEX productAttrName_lower ON ProductTypeTransactionAttributes(lower(attrName));
+
+	-- milk
+	INSERT INTO ProductTypeTransactionAttributes VALUES(0, 0, 'milkDensityGramsPerMillilitre');
+
+-- TransactionAttributes
+	-- the values for the unique attributes for every transaction
+
+CREATE TABLE ProductTransactionAttribute (
+	productTransactionUuid UUID REFERENCES ProductTransactions(productTransactionUuid) NOT NULL,
+	attrId SERIAL REFERENCES ProductTypeTransactionAttributes(attrId) NOT NULL,
+	attrValue VARCHAR(255) NOT NULL,
+	UNIQUE(productTransactionUuid, attrId) -- for every transaction, all of its attributes must only have one value
+);
+
+-- Loans
+
+CREATE TABLE Loans (
+	loanUuid UUID PRIMARY KEY NOT NULL DEFAULT uuid_generate_v4(),
+	moneyTransactionUuid UUID REFERENCES MoneyTransactions(moneyTransactionUuid) NOT NULL,
+	dueDate DATE
```

Comment:
how come sometimes we use date and sometimes we use timestamp? do you think a loan due date should have a time associated?

Reply:
No. Everything by default is timestamp. I just don't feel like a time stamp is common at all for loan due dates.

Reply:
kk, so we when we figure out the weekly payment we can just do the math from the initial loan transaction date and due date (disregarding time)